llvm/lib/Target/EVM/EVMStackSolver.cpp

```diff
@@ -202,75 +202,73 @@ void EVMStackSolver::run() {
   });
 }
 
-Stack EVMStackSolver::propagateStackThroughInst(const Stack &AfterInst,
-                                                const Operation &Op,
-                                                bool CompressStack) {
+Stack EVMStackSolver::propagateStackThroughMI(const Stack &AfterMI,
+                                              const MachineInstr &MI,
+                                              bool CompressStack) {
   // Enable aggressive stack compression for recursive calls.
-  if (Op.isFunctionCall())
+  if (MI.getOpcode() == EVM::FCALL)
     // TODO: compress stack for recursive functions.
     CompressStack = false;
 
-  const SmallVector<StackSlot *> InstDefs =
-      StackModel.getSlotsForInstructionDefs(Op.getMachineInstr());
+  const SmallVector<StackSlot *> MIDefs =
+      StackModel.getSlotsForInstructionDefs(&MI);
 
-  // Determine the ideal permutation of the slots in ExitLayout that are not
-  // operation outputs (and not to be generated on the fly), s.t. shuffling the
-  // 'IdealStack + Operation.output' to ExitLayout is cheap.
-  Stack BeforeInst = calculateStackBeforeInst(
-      InstDefs, AfterInst, CompressStack, StackModel.stackDepthLimit());
+  // Determine the ideal permutation of the slots in AfterMI that are not
+  // MI defs, s.t. shuffling the 'BeforeMI + MIDefs' to AfterMI is cheap.
+  Stack BeforeMI = calculateStackBeforeInst(MIDefs, AfterMI, CompressStack,
+                                            StackModel.stackDepthLimit());
 
 #ifndef NDEBUG
   // Make sure the resulting previous slots do not overlap with any assigned
   // variables.
-  if (Op.isAssignment())
-    for (auto *StackSlot : BeforeInst)
+  if (isConstCopyOrLinkerMI(MI))
```

Comment:
Could you, please clarify why do we need `isConstCopyOrLinkerMI(MI)` constrain? Before the change, `Op.isAssignment()` was created for each MI.

Reply:
Removed.